src/main/java/appeng/container/implementations/ContainerFluidTerminal.java

```diff
@@ -0,0 +1,479 @@
+	{
+		super( ip, terminal );
+		this.terminal = terminal;
+		this.clientCM = new ConfigManager( this );
+
+		this.clientCM.registerSetting( Settings.SORT_BY, SortOrder.NAME );
+		this.clientCM.registerSetting( Settings.SORT_DIRECTION, SortDir.ASCENDING );
+		this.clientCM.registerSetting( Settings.VIEW_MODE, ViewItems.ALL );
+		if( Platform.isServer() )
+		{
+			this.serverCM = terminal.getConfigManager();
+			this.monitor = terminal.getInventory( AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) );
+			if( this.monitor != null )
+			{
+				this.monitor.addListener( this, null );
+				final IGridNode node = terminal.getGridNode( AEPartLocation.INTERNAL );
+				if( node != null )
+				{
+					this.networkNode = node;
+					final IGrid g = node.getGrid();
+					if( g != null )
+					{
+						this.setPowerSource( new ChannelPowerSrc( this.networkNode, g.getCache( IEnergyGrid.class ) ) );
+					}
+				}
+			}
+		}
+		else
+		{
+			this.monitor = null;
+		}
+		this.bindPlayerInventory( ip, 0, 222 - 82 );
+	}
+
+	@Override
+	public boolean isValid( Object verificationToken )
+	{
+		return true;
+	}
+
+	@Override
+	public void postChange( IBaseMonitor<IAEFluidStack> monitor, Iterable<IAEFluidStack> change, IActionSource actionSource )
+	{
+		for( final IAEFluidStack is : change )
+		{
+			this.fluids.add( is );
+		}
+	}
+
+	@Override
+	public void addListener( IContainerListener listener )
+	{
+		super.addListener( listener );
+
+		this.queueInventory( listener );
+	}
+
+	@Override
+	public void onListUpdate()
+	{
+		for( final IContainerListener c : this.listeners )
+		{
+			this.queueInventory( c );
+		}
+	}
+
+	private void queueInventory( final IContainerListener c )
+	{
+		if( Platform.isServer() && c instanceof EntityPlayer && this.monitor != null )
+		{
+			try
+			{
+				PacketMEFluidInventoryUpdate piu = new PacketMEFluidInventoryUpdate();
+				final IItemList<IAEFluidStack> monitorCache = this.monitor.getStorageList();
+
+				for( final IAEFluidStack send : monitorCache )
+				{
+					try
+					{
+						piu.appendFluid( send );
+					}
+					catch( final BufferOverflowException boe )
+					{
+						NetworkHandler.instance().sendTo( piu, (EntityPlayerMP) c );
+
+						piu = new PacketMEFluidInventoryUpdate();
+						piu.appendFluid( send );
+					}
+				}
+
+				NetworkHandler.instance().sendTo( piu, (EntityPlayerMP) c );
+			}
+			catch( final IOException e )
+			{
+				AELog.debug( e );
+			}
+		}
+	}
+
+	@Override
+	public IConfigManager getConfigManager()
+	{
+		if( Platform.isServer() )
+		{
+			return this.serverCM;
+		}
+		return this.clientCM;
+	}
+
+	@Override
+	public void onContainerClosed( final EntityPlayer player )
+	{
+		super.onContainerClosed( player );
+		if( this.monitor != null )
+		{
+			this.monitor.removeListener( this );
+		}
+	}
+
+	@Override
+	public void doAction( EntityPlayerMP player, InventoryAction action, int slot, long id )

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') The Cyclomatic Complexity of this method "doAction" is 29 which is greater than 10 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)